synapse/appservice/api.py

```diff
@@ -444,6 +444,44 @@ async def claim_client_keys(
 
         return response, missing
 
+    async def query_keys(
+        self, service: "ApplicationService", query: Dict[str, List[str]]
+    ) -> Dict[str, Dict[str, Dict[str, JsonDict]]]:
+        """Query the application service for keys.
+
+        Args:
+            query: An iterable of tuples of (user ID, device ID, algorithm).
+
+        Returns:
+            A map of device_keys/master_keys/self_signing_keys/user_signing_keys:
+
+            device_keys is a map of user ID -> a map device ID -> device info.
+        """
+        if service.url is None:
+            return {}
+
+        # This is required by the configuration.
+        assert service.hs_token is not None
+
+        uri = f"{service.url}/_matrix/app/unstable/org.matrix.msc3984/keys/query"
+        try:
+            response = await self.post_json_get_json(
+                uri,
+                query,
+                headers={"Authorization": [f"Bearer {service.hs_token}"]},
+            )
+        except CodeMessageException as e:
+            # The appservice doesn't support this endpoint.
+            if e.code == 404 or e.code == 405:
+                return {}
+            logger.warning("claim_keys to %s received %s", uri, e.code)
+            return {}
+        except Exception as ex:
+            logger.warning("claim_keys to %s threw exception %s", uri, ex)
+            return {}
```

Comment:
is it really OK to just eat up errors and return `{}`? It's not clear how come this is safe — would benefit from an explanation here/in docstring.

Reply:
The MSC says:

> If the appservice responds with an error of any kind (including timeout), the homeserver should treat the appservice's response as `{}`.

I can expand the docstirng though.

Reply:
Please leave a comment on the MSC!

Reply:
Cross-link https://github.com/matrix-org/matrix-spec-proposals/pull/3984#discussion_r1152102625

Are you OK with closing this for now then? We can update the implementation based on changes to the MSC.

Reply:
It'd be good to get something into the code because realistically as soon as this thread gets closed it will be forgotten about for the rest of time and we'll be stuck with it.

Unfortunately I think the answer is slowly revealing itself as 'this isn't safe to do but we do it anyway'. This feels like a design problem in an already warty system — I'm not really happy with it and it feels like we should consider whether there are other options here.

Reply:
Well we have the thread in the MSC for tracking. That's usually what we've done for this sort of thing.